tests/test_grid.py

```diff
@@ -6,103 +6,86 @@
 DESCRIPTION
 This module tests the following functions from grid.py:
 
-- class CoreGridConfig:
-    Configure the `core.grid` module.
-    This data class is used to setup the arrangement of grid cells to be used in
-    running a `virtual_rainforest` simulation.
-
-- CoreGrid._make_square_grid(cell_area: float, cell_nx: int, cell_ny: int) -> dict:
+- grid.make_square_grid
     Create a square grid.
 
-- CoreGrid._make_hex_grid(cell_area: float, cell_nx: int, cell_ny: int) -> dict:
+- grid.make_hex_grid
     Create a hexagonal grid.
 
-- TODO CoreGrid.export geojson
-    Export grid system as geojson file.
-    outfile: A self.cell_dict returned by a make_square_grid or make_hex_grid call.
+- TODO The grid.Grid class.
 
-- TODO Once it is extracted from where it is, _feature.
+- TODO grid.Grid.dump and grid.Grid.dumps for0 geojson
 
 """
 
+import numpy as np
+import numpy.linalg as LA
 import pytest
 from hypothesis import given, settings
 from hypothesis.strategies import integers
+from scipy.spatial.distance import euclidean  # type: ignore
```

Comment:
That's interesting.  I like the clarity of seeing which bits of the package are being tested at the top of the file, but I think I can see a point for imports from the _package being tested_ inside the test function. If the import doesn't work then only the test using that import fails rather than the whole test file and the test is effectively also testing that the objects import correctly. However that does seem like quite a fundamental problem - if some imports are broken, the rest of the file completing is a bit secondary!

 For other imports, having them once at the top of the file seems cleaner?

Reply:
I tend to import everything but pytest-spcecific stuff (and hypothesis) within each test to make sure everything is reset as much as possible in between tests. It makes it clearer which tests needs what package. And finally, also, facilitates mocking, as you cannot mock something if is already imported.

Having said that, I think it is fine to import general tools like numpy or scipy at the top.

Reply:
Makes sense. There's a bit of a grey line about which goes where and we should standardise that, but I can clearly see the argument for imports from within the project.